src/plugins/syscalls/linux.cpp

```diff
@@ -159,193 +183,213 @@ static std::vector<uint64_t> linux_build_argbuf(vmi_instance_t vmi,
+        keyval("Syscall", fmt::Nval((uint64_t)(params->num))),
+        keyval("NArgs", fmt::Nval(params->sc->num_args)),
+        keyval("Type", fmt::Estr(params->type)),
+        this->fmt_args
+    );
+}
 
-    print_sysret(s, drakvuf, info, w->num, info->trap->breakpoint.module, sc, info->regs->rax);
+event_response_t linux_syscalls::linux_ret_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    auto params = libhook::GetTrapParams<linux_syscall_data>(info);
+    if (!drakvuf_check_return_context(drakvuf, info, params->pid, params->tid, params->rsp))
+        return VMI_EVENT_RESPONSE_NONE;
 
-    drakvuf_remove_trap(drakvuf, info->trap, (drakvuf_trap_free_t)free_trap);
-    s->traps = g_slist_remove(s->traps, info->trap);
+    this->print_sysret(drakvuf, info, (uint64_t)params->num);
 
-    return 0;
+    auto hookID = make_hook_id(info);
+    this->ret_hooks.erase(hookID);
+
+    return VMI_EVENT_RESPONSE_NONE;
 }
 
-static event_response_t linux_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+event_response_t linux_syscalls::linux_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
 {
-    auto vmi = vmi_lock_guard(drakvuf);
-    struct wrapper* w = (struct wrapper*)info->trap->data;
-    syscalls* s = w->s;
-
-    const syscall_t* sc = NULL;
-    addr_t pt_regs = 0;
-
+    addr_t pt_regs_addr = 0;
     addr_t nr = ~0;
-    if ( VMI_GET_BIT(info->regs->rdi, 47) )
-    {
-        /*
-         * On older kernels: __visible void do_syscall_64(struct pt_regs *regs)
-         */
-        pt_regs = info->regs->rdi;
-        (void)vmi_read_addr_va(vmi, pt_regs + s->offsets[PT_REGS_ORIG_RAX], 0, &nr);
-    }
-    else
-    {
-        /*
-         * On newer kernels: __visible void do_syscall_64(unsigned long nr, struct pt_regs *regs)
-         */
-        nr = info->regs->rdi;
-        pt_regs = info->regs->rsi;
-    }
 
-    if ( nr<NUM_SYSCALLS_LINUX )
+    if (!get_pt_regs_addr(drakvuf, info, &pt_regs_addr, &nr))
     {
-        sc = linuxsc::linux_syscalls[nr];
-
-        if ( s->filter && !g_hash_table_contains(s->filter, sc->name) )
-            return 0;
+        PRINT_DEBUG("[SYSCALLS] Failed to get pt_regs_addr in %s\n", info->trap->name);
+        return VMI_EVENT_RESPONSE_NONE;
     }
 
-    auto args = linux_build_argbuf(vmi, info, s, sc, pt_regs);
+    auto arguments = build_arguments_buffer(drakvuf, info, pt_regs_addr, nr);
+    this->print_syscall(drakvuf, info, arguments);
 
-    print_syscall(s, drakvuf, info, nr, info->trap->breakpoint.module, sc, args, false);
+    if (this->disable_sysret)
+        return VMI_EVENT_RESPONSE_NONE;
 
-    if ( s->disable_sysret )
-        return 0;
+    auto hook = this->createReturnHook<linux_syscall_data>(info, &linux_syscalls::linux_ret_cb);
+    auto params = libhook::GetTrapParams<linux_syscall_data>(hook->trap_);
+    params->rsp = drakvuf_get_function_return_address(drakvuf, info);
+    params->pid = info->proc_data.pid;
+    params->tid = info->proc_data.tid;
```

Comment:
There's already a helper for this.
```suggestion
    params->setResultCallParams(info);
```